File: packages/offline-contract/receipt-parser.mjs
import { categories, exactKeys, limits, parseAmount, requireThat as check, validUnicode } from './contract.mjs';

const horizontal = '[ \\t\\u00a0\\u202f]';
const currencyMarker = '(?:CAD(?:[ \\t]*\\$)?|CA\\$|C\\$|\\$)';
const suffixMarker = '(?:\\$[ \\t]*CAD|CAD|CA\\$|C\\$|\\$)';
const merchantLabel = /^(?:merchant|vendor|marchand|commerçant)\s*:\s*(.*)$/iu;
const ignoredTotal = /^(?:sub[ -]?total|sous[ -]?total|total\s+(?:tax(?:es)?|items?|articles?|discounts?|remises?)|tax(?:es)?|tps|tvq|hst|gst|pst|tender(?:ed)?|cash|change|discount|remise|monnaie|comptant)\b/iu;
const instructionLike = /\b(?:ignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions|system\s+prompt|assistant\s*:|execute\s+(?:code|command)|send\s+(?:money|data)|ignorez\s+les\s+instructions)(?!\p{L})/iu;
const foreignCurrency = /\b(?:USD|EUR|GBP|AUD|NZD|JPY|CNY|INR|MXN|CHF|RMB|US\s+dollars?)(?![a-z])|US\s*\$|[€£¥₹]/iu;
const refund = /\b(?:refund(?:ed)?|remboursement|remboursé|credit\s+note|note\s+de\s+crédit)(?!\p{L})/iu;
const signed = /[-−+()]|\bCR\b/iu;
function input(text, locale) {
  check(locale === 'en-CA' || locale === 'fr-CA', 'unsupported_locale');
  check(validUnicode(text), 'invalid_unicode');
  check([...text].length <= 4000, 'input_too_long');
}
function validMerchant(value) {
  return value.length > 0 && [...value].length <= 200 && /\p{L}/u.test(value) && !/[\p{Cc}]/u.test(value)
    && !/^(?:receipt|invoice|facture|welcome|thank you|merci)$/iu.test(value)
    && !/\d+[.,]\d{2}/u.test(value) && !ignoredTotal.test(value);
}
function totalLabel(locale) {
  return locale === 'en-CA' ? /^(?:grand\s+total|total\s+paid|amount\s+due|balance\s+due|total)\b\s*:?\s*(.*)$/iu
    : /^(?:total\s+général|montant\s+total|total\s+à\s+payer|net\s+à\s+payer|total\s+ttc|total)\b\s*:?\s*(.*)$/iu;
}
function amount(text, locale) {
  check(!signed.test(text), 'signed_total');
  const numeric = locale === 'en-CA' ? '(?:0|[1-9][0-9]*|[1-9][0-9]{0,2}(?:,[0-9]{3})+)\\.[0-9]{2}'
    : `(?:0|[1-9][0-9]*|[1-9][0-9]{0,2}(?:${horizontal}[0-9]{3})+),[0-9]{2}`;
  const match = new RegExp(`^(?:${currencyMarker}${horizontal}*)?(${numeric})(?:${horizontal}*${suffixMarker})?$`, 'iu').exec(text);
  check(match, 'malformed_total');
  const normalized = locale === 'en-CA' ? match[1].replaceAll(',', '') : match[1].replace(/[ \t\u00a0\u202f]/gu, '').replace(',', '.');
  check(normalized !== '0.00', 'nonpositive_total');
  try { return parseAmount(normalized); } catch { throw new Error('total_overflow'); }
}

/** Suggestions only. This module has no persistence, model, network or tool path. */
export function parseReceipt(sourceText, locale) {
  input(sourceText, locale);
  const proposal = { parserVersion: 1, locale, sourceText, merchant: null, amountMinor: null, currencyCode: null, merchantLineIndex: null, totalLineIndex: null, category: null, requiresReview: true, reasons: [] };
  const lines = sourceText.split(/\r\n|\r|\n/u).map((raw, index) => ({ text: raw.trim(), index })).filter(line => line.text !== '');
  if (instructionLike.test(sourceText)) { proposal.reasons.push('instruction_like_text'); return proposal; }
  const labelled = lines.map(line => ({ ...line, match: merchantLabel.exec(line.text) })).filter(line => line.match);
  if (labelled.length > 1) proposal.reasons.push('ambiguous_merchant');
  else {
    const candidate = labelled.length ? { text: labelled[0].match[1].trim(), index: labelled[0].index } : lines[0];
    if (candidate && (labelled.length || !totalLabel(locale).test(candidate.text)) && validMerchant(candidate.text)) { proposal.merchant = candidate.text; proposal.merchantLineIndex = candidate.index; }
    else proposal.reasons.push('missing_merchant');
  }
  if (foreignCurrency.test(sourceText)) { proposal.reasons.push('foreign_currency'); return proposal; }
  if (refund.test(sourceText)) { proposal.reasons.push('unsupported_refund'); return proposal; }
  const totals = lines.filter(line => !merchantLabel.test(line.text) && !ignoredTotal.test(line.text)).map(line => ({ ...line, match: totalLabel(locale).exec(line.text) })).filter(line => line.match);
  if (totals.length === 0) proposal.reasons.push('missing_total');
  else if (totals.length > 1) proposal.reasons.push('multiple_totals');
  else {
    try {
      proposal.amountMinor = amount(totals[0].match[1].trim(), locale); proposal.totalLineIndex = totals[0].index; proposal.currencyCode = 'CAD';
      if (!/\bCAD(?![a-z])|CA\$|C\$/iu.test(sourceText)) proposal.reasons.push('currency_assumed_cad');
    } catch (error) { proposal.reasons.push(error.message); }
  }
  return proposal;
}

/** Optional on-device models may classify a grounded proposal, never invent cash. */
export function validateModelProposal(sourceText, locale, output) {
  const proposal = parseReceipt(sourceText, locale);
  exactKeys(output, ['merchant', 'amountMinor', 'currencyCode', 'category']);
  check(output.merchant === null || (typeof output.merchant === 'string' && output.merchant === proposal.merchant), 'model_merchant_ungrounded');
  check(output.amountMinor === null || (Number.isSafeInteger(output.amountMinor) && output.amountMinor > 0 && output.amountMinor <= limits.amountMinor && output.amountMinor === proposal.amountMinor), 'model_amount_ungrounded');
  check(output.currencyCode === (output.amountMinor === null ? null : 'CAD'), 'model_currency_ungrounded');
  check(output.category === null || categories.includes(output.category), 'model_category_invalid');
  check(output.category === null || (proposal.merchant !== null && proposal.amountMinor !== null), 'model_category_ungrounded');
  return { ...proposal, category: output.category };
}

File: apps/android/app/src/main/java/ca/penny/offline/ReceiptParser.kt
package ca.penny.offline

import org.json.JSONArray
import org.json.JSONObject

/** Exact capture-v1 proposal. Source text is immutable evidence, never instructions. */
data class ReceiptDraft(val sourceText: String, val locale: String, val merchant: String? = null, val amountMinor: Long? = null,
    val merchantLineIndex: Int? = null, val totalLineIndex: Int? = null, val category: String? = null, val reasons: List<String> = emptyList()) {
    val amount: String get() = amountMinor?.let(Money::edit).orEmpty()
    fun json() = JSONObject().put("parserVersion",1).put("locale",locale).put("sourceText",sourceText)
        .put("merchant",merchant ?: JSONObject.NULL).put("amountMinor",amountMinor ?: JSONObject.NULL)
        .put("currencyCode",if(amountMinor == null) JSONObject.NULL else "CAD")
        .put("merchantLineIndex",merchantLineIndex ?: JSONObject.NULL).put("totalLineIndex",totalLineIndex ?: JSONObject.NULL)
        .put("category",category ?: JSONObject.NULL).put("requiresReview",true).put("reasons",JSONArray(reasons))
}
object ReceiptParser {
    // Java's UNICODE_CASE folds Turkish dotted/dotless I into ASCII I; ECMAScript /iu does not.
    // Match ASCII case plus the simple folds used by these labels, retaining original capture spans.
    private class Rule(pattern: String) {
        private val compiled=java.util.regex.Pattern.compile(pattern,java.util.regex.Pattern.CASE_INSENSITIVE)
        private fun matcher(value: String)=compiled.matcher(value.replace('ſ','s').replace('K','k'))
        fun matches(value: String)=matcher(value).matches()
        fun containsMatchIn(value: String)=matcher(value).find()
        fun matchEntire(value: String): Groups? {
            val match=matcher(value); if(!match.matches()) return null
            return Groups((0..match.groupCount()).map { if(match.start(it)<0) "" else value.substring(match.start(it),match.end(it)) })
        }
    }
    private data class Groups(val groupValues: List<String>)
    private fun regex(pattern: String) = Rule(pattern.replace("\\s", "[\\u0009-\\u000d \\u00a0\\u1680\\u2000-\\u200a\\u2028\\u2029\\u202f\\u205f\\u3000\\ufeff]")
        .replace("\\b", "(?:(?<=[A-Za-z0-9_])(?![A-Za-z0-9_])|(?<![A-Za-z0-9_])(?=[A-Za-z0-9_]))")
        .replace("é","[éÉ]").replace("ç","[çÇ]").replace("à","[àÀ]"))
    private val merchantLabel = regex("""^(?:merchant|vendor|marchand|commerçant)\s*:\s*(.*)$""")
    private val ignored = regex("""^(?:sub[ -]?total|sous[ -]?total|total\s+(?:tax(?:es)?|items?|articles?|discounts?|remises?)|tax(?:es)?|tps|tvq|hst|gst|pst|tender(?:ed)?|cash|change|discount|remise|monnaie|comptant)\b""")
    private val instruction = regex("""\b(?:ignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions|system\s+prompt|assistant\s*:|execute\s+(?:code|command)|send\s+(?:money|data)|ignorez\s+les\s+instructions)(?!\p{L})""")
    private val foreign = regex("""\b(?:USD|EUR|GBP|AUD|NZD|JPY|CNY|INR|MXN|CHF|RMB|US\s+dollars?)(?![a-z])|US\s*\$|[€£¥₹]""")
    private val refund = regex("""\b(?:refund(?:ed)?|remboursement|remboursé|credit\s+note|note\s+de\s+crédit)(?!\p{L})""")
    private val signed = regex("""[-−+()]|\bCR\b""")
    private fun total(locale: String) = regex(if(locale == "en-CA") """^(?:grand\s+total|total\s+paid|amount\s+due|balance\s+due|total)\b\s*:?\s*(.*)$""" else """^(?:total\s+général|montant\s+total|total\s+à\s+payer|net\s+à\s+payer|total\s+ttc|total)\b\s*:?\s*(.*)$""")
    private fun validMerchant(value: String) = value.isNotEmpty() && value.codePointCount(0,value.length) <= 200 && Regex("\\p{L}").containsMatchIn(value) && !Regex("\\p{Cc}").containsMatchIn(value) && !regex("^(?:receipt|invoice|facture|welcome|thank you|merci)$").matches(value) && !Regex("[0-9]+[.,][0-9]{2}").containsMatchIn(value) && !ignored.containsMatchIn(value)
    private fun amount(text: String, locale: String): Long {
        require(!signed.containsMatchIn(text)) { "signed_total" }
        val horizontal = "[ \\t\\u00a0\\u202f]"
        val numeric = if(locale == "en-CA") "(?:0|[1-9][0-9]*|[1-9][0-9]{0,2}(?:,[0-9]{3})+)\\.[0-9]{2}" else "(?:0|[1-9][0-9]*|[1-9][0-9]{0,2}(?:${horizontal}[0-9]{3})+),[0-9]{2}"
        val match = regex("^(?:(?:CAD(?:[ \\t]*\\$)?|CA\\$|C\\$|\\$)${horizontal}*)?($numeric)(?:${horizontal}*(?:\\$[ \\t]*CAD|CAD|CA\\$|C\\$|\\$))?$").matchEntire(text)
        requireNotNull(match) { "malformed_total" }
        val normalized = if(locale == "en-CA") match.groupValues[1].replace(",", "") else match.groupValues[1].replace(Regex(horizontal), "").replace(',', '.')
        require(normalized != "0.00") { "nonpositive_total" }
        return try { Money.parse(normalized) } catch(_: Exception) { error("total_overflow") }
    }
    fun draft(sourceText: String, locale: String = "en-CA"): ReceiptDraft {
        require(locale in listOf("en-CA", "fr-CA")) { "unsupported_locale" }
        require(runCatching { Wire.requireUnicode(sourceText) }.isSuccess) { "invalid_unicode" }
        require(sourceText.codePointCount(0,sourceText.length) <= 4000) { "input_too_long" }
        var result = ReceiptDraft(sourceText,locale)
        val reasons = mutableListOf<String>()
        val lines = sourceText.split(Regex("\r\n|\r|\n")).mapIndexed { index, line -> index to Wire.trim(line) }.filter { it.second.isNotEmpty() }
        if(instruction.containsMatchIn(sourceText)) return result.copy(reasons=listOf("instruction_like_text"))
        val labelled = lines.filter { merchantLabel.matches(it.second) }
        if(labelled.size > 1) reasons += "ambiguous_merchant" else {
            val candidate = labelled.firstOrNull()?.let { it.first to Wire.trim(merchantLabel.matchEntire(it.second)!!.groupValues[1]) } ?: lines.firstOrNull()
            if(candidate != null && (labelled.isNotEmpty() || !total(locale).matches(candidate.second)) && validMerchant(candidate.second)) result = result.copy(merchant=candidate.second,merchantLineIndex=candidate.first)
            else reasons += "missing_merchant"
        }
        if(foreign.containsMatchIn(sourceText)) return result.copy(reasons=reasons+"foreign_currency")
        if(refund.containsMatchIn(sourceText)) return result.copy(reasons=reasons+"unsupported_refund")
        val totals = lines.filter { !merchantLabel.matches(it.second) && !ignored.containsMatchIn(it.second) && total(locale).matches(it.second) }
        when {
            totals.isEmpty() -> reasons += "missing_total"
            totals.size > 1 -> reasons += "multiple_totals"
            else -> try {
                result = result.copy(amountMinor=amount(Wire.trim(total(locale).matchEntire(totals[0].second)!!.groupValues[1]),locale),totalLineIndex=totals[0].first)
                if(!regex("""\bCAD(?![a-z])|CA\$|C\$""").containsMatchIn(sourceText)) reasons += "currency_assumed_cad"
            } catch(e: Exception) { reasons += checkNotNull(e.message) }
        }
        return result.copy(reasons=reasons)
    }
    fun validateModel(source: String, locale: String, output: JSONObject): ReceiptDraft {
        val draft = draft(source,locale)
        Wire.exactKeys(output,"merchant","amountMinor","currencyCode","category")
        require(output.isNull("merchant") || output.get("merchant") is String && output.getString("merchant") == draft.merchant) { "model_merchant_ungrounded" }
        require(output.isNull("amountMinor") || runCatching { Wire.integer(output,"amountMinor").let { it > 0 && it == draft.amountMinor } }.getOrDefault(false)) { "model_amount_ungrounded" }
        require(if(output.isNull("amountMinor")) output.isNull("currencyCode") else output.get("currencyCode") == "CAD") { "model_currency_ungrounded" }
        require(output.isNull("category") || output.get("category") in Categories.all) { "model_category_invalid" }
        require(output.isNull("category") || draft.merchant != null && draft.amountMinor != null) { "model_category_ungrounded" }
        return draft.copy(category=if(output.isNull("category")) null else output.getString("category"))
    }
    fun reason(value: String) = when(value) {
        "currency_assumed_cad" -> "CAD is assumed from the selected locale. Confirm the currency."
        "foreign_currency" -> "A foreign currency was found. Enter a verified CAD amount manually."
        "multiple_totals" -> "More than one total was found. Choose the correct amount manually."
        "instruction_like_text" -> "Instruction-like text was ignored. Enter the expense manually."
        "input_too_long" -> "OCR text exceeds 4,000 characters. No fields were proposed; the full text is available."
        "ocr_failed" -> "Text could not be read. The image is retained for manual entry."
        else -> value.replace('_',' ').replaceFirstChar { it.uppercase() } + ". Review the original receipt."
    }
}
